---
# 1.2.1 NumPy Basics
---

`NumPy`, which stands for **Numerical Python**, is a cornerstone of scientific computing in Python. 

It allows you to work with large, multi-dimensional arrays and matrices, and it provides a plethora of mathematical functions to manipulate these arrays efficiently. 

Whether you’re doing data analysis, machine learning, or any form of computational science, NumPy is a must-know tool because of its speed and versatility.


---
## 1.2.1.1  Creating NumPy Arrays 
---

`NumPy` provides various ways to create arrays. 

You can convert lists or tuples into arrays using `np.array()`. 

Additionally, you can create arrays filled with 
*  zeros using `np.zeros()`, 
*  ones using `np.ones()`, 
*  evenly spaced values using `np.arange()`, and 
*  linearly spaced values using `np.linspace()`. 

Let’s look at some examples.

In [1]:
## Code snippet 1.2.1.1

import numpy as np

array_from_list = np.array([1, 2, 3])
array_of_zeros = np.zeros((2, 3))
array_of_ones = np.ones((2, 3))
array_range = np.arange(0, 10, 2)
array_linspace = np.linspace(0, 1, 5)

print("Array from list:", array_from_list)
print("Array of zeros:", array_of_zeros)
print("Array of ones:", array_of_ones)
print("Array with range:", array_range)
print("Array with linspace:", array_linspace)


Array from list: [1 2 3]
Array of zeros: [[0. 0. 0.]
 [0. 0. 0.]]
Array of ones: [[1. 1. 1.]
 [1. 1. 1.]]
Array with range: [0 2 4 6 8]
Array with linspace: [0.   0.25 0.5  0.75 1.  ]


---
## 1.2.1.2 Array Attributes
---

Once you have created an array, it’s important to understand its structure and properties. 

NumPy arrays have several attributes that give you this information. 
*  `shape` tells you the dimensions of the array, 
*  `size` gives the total number of elements, 
*  `dtype` shows the data type of the elements, and 
*  `ndim` provides the number of dimensions of the array. 

Let’s explore these attributes.

In [5]:
## Code snippet 1.2.1.2

array = np.array([[1, 2, 3], [4, 5, 6]])

print(array,"\n")
print("Shape:", array.shape)
print("Size:", array.size)
print("Data type:", array.dtype)
print("Number of dimensions:", array.ndim)

[[1 2 3]
 [4 5 6]] 

Shape: (2, 3)
Size: 6
Data type: int64
Number of dimensions: 2


---
## 1.2.1.3  Array Indexing and Slicing
---

Accessing elements in a NumPy array is similar to lists in Python but also extends to multi-dimensional arrays. 

You can use *indexing* to access individual elements and *slicing* to extract subarrays. 

Let’s see how it works with some examples.

In [10]:
## Code snippet 1.2.1.3

array = np.array([[1, 2, 3], [4, 5, 6]])

print(array,"\n")
print("First element:", array[0])
print("Element at (0,1):", array[0, 1])
print("Second row:", array[1, :])
print("Third column:", array[:, 2])

[[1 2 3]
 [4 5 6]] 

First element: [1 2 3]
Element at (0,1): 2
Second row: [4 5 6]
Third column: [3 6]


---
##  1.2.1.4  Array operations
---

`NumPy` makes it easy to perform element-wise arithmetic operations on arrays. 

You can *add*, *subtract*, *multiply*, and *divide* arrays directly. 

Additionally, `NumPy` supports *broadcasting*, which allows you to perform operations on arrays of different shapes. 

Here are some examples.

In [26]:
## Code snippet 1.2.1.4a

array1 = np.array([1, 2, 3])
array2 = np.array([4, 5, 6])

print("array1 = ", array1)
print("array2 = ", array2,"\n")
print("Addition:", array1 + array2)
print("Multiplication by scalar:", array1 * 2)
print("Division:", array1 / array2)


array1 =  [1 2 3]
array2 =  [4 5 6] 

Addition: [5 7 9]
Multiplication by scalar: [2 4 6]
Division: [0.25 0.4  0.5 ]


In [28]:
## Code snippet 1.2.1.4b
# Broadcasting

A = np.array([[1, 2, 3], [4, 5, 6]])
B = np.array([[3],[6]])

print(f'A = \n {A}')
print(f'B = \n {B}\n')
print(f'A + B = \n {A + B}')


A = 
 [[1 2 3]
 [4 5 6]]
B = 
 [[3]
 [6]]

A + B = 
 [[ 4  5  6]
 [10 11 12]]


---
## 1.2.1.5  Universal Functions (ufuncs)
---

`NumPy` provides a set of universal functions (`ufuncs`) that perform element-wise operations on arrays. 

These functions are highly optimized and provide a convenient way to apply mathematical operations to entire arrays. 

Let’s look at a few common ufuncs.

In [36]:
## Code snippet 1.2.1.5

array = np.array([1, 2, 3, 4])

print("Square root:", np.sqrt(array))
print("Exponential:", np.exp(array))
print("Sine:", np.sin(array))


Square root: [1.         1.41421356 1.73205081 2.        ]
Exponential: [ 2.71828183  7.3890561  20.08553692 54.59815003]
Sine: [ 0.84147098  0.90929743  0.14112001 -0.7568025 ]


*  `np.sqrt(array)` : 

    - $[\sqrt{1} \;\;\sqrt{2} \;\; \sqrt{3} \;\;\;\sqrt{4}] =
  \texttt{[1.         1.41421356 1.73205081 2.]}$

*  `np.exp(array)` :

    - $[e^{1}\;\; e^{2}\;\; e^{3}\;\; e^{4}] = 
  \texttt{[ 2.71828183  7.3890561  20.08553692 54.59815003]}$

*  `np.sin(array)` :

    - $[\sin(1) \;\; \sin(2) \;\; \sin(3) \;\; \sin(4)] = 
    \texttt{[ 0.84147098  0.90929743  0.14112001 -0.7568025 ]}$

---
## 1.2.1.6 Basic Statistical Operations in NumPy
---

`NumPy` also provides a range of statistical functions that are very useful for data analysis. 

You can calculate the `mean`, `median`, `standard deviation`, and `sum` of array elements efficiently. 

These functions help in understanding the data distribution and variability. Let’s see some examples.


In [37]:
## Code snippet 1.2.16

array = np.array([1, 2, 3, 4, 5])

print("Mean:", np.mean(array))
print("Median:", np.median(array))
print("Standard Deviation:", np.std(array))
print("Sum:", np.sum(array))

Mean: 3.0
Median: 3.0
Standard Deviation: 1.4142135623730951
Sum: 15


---
##  Summary
---

This lecture introduces the basics of NumPy, including 

*  creating arrays, 
*  understanding array attributes, 
*  indexing and slicing arrays, 
*  performing array operations, 
*  using universal functions, and
*  conducting basic statistical operations.

These fundamental concepts are essential for effective scientific computing and data analysis in Python.
